### PR TITLE
Changes to setting up SSH authentication for Git

### PR DESCRIPTION
**Pull request by OctoElla**
# Changes
Modified the "Configuring `ssh-auth` authentication for Git" section of the "Authentication at Run Time" doc. 
According to the issue: Make it easier to set up SSH authentication for Git #3631

/kind documentation

 # Submitter Checklist
 As the author of this PR, please check off the items in this checklist:
 
 * [x]  [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
 * [ ]  [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
 * [x]  Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
 * [x]  Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
   functionality, content, code)
 * [x]  Release notes block below has been filled in or deleted (only if no user facing changes)
# Release Notes

```release-note
NONE
```

### CODE REVIEW

**Comment by pritidesai** [comment]
thank you @OctoElla for this! Welcome to the community 🎉 
/ok-to-test

File: docs/auth.md [hunk 1]
```diff
@@ -235,88 +235,41 @@ Note: Github deprecated basic authentication with username and password. You can
...
-   ```yaml
-   apiVersion: v1
-   kind: ServiceAccount
-   metadata:
-     name: build-bot
-   secrets:
-     - name: ssh-key
-   ```
-
-1. In `run.yaml`, associate the `ServiceAccount` with your `Run` by doing one of the following:
```
Comment (pritidesai) [inline docs/auth.md:286]: thank you @OctoElla for updating the instructions!

Can we please keep this part of the doc with the new instructions?

**Comment by OctoElla** [comment]
@sbwsg 
I’ve tested new steps from the instruction.
this step is not working:
```
4. Create a secret from the file with the private key:
        kubectl create secret generic github-private-key --from-file=id_rsa
```
If I use this command I get a secret with the type:
```
type: Opaque
```
But drawing a conclusion from my experiment, we need the secret with the type:
```
type: kubernetes.io/ssh-auth
```
and an annotation:
```
  annotations:
    tekton.dev/git-0: github.com 
```
Only in this case, Tekton creates a ~/.ssh/config file containing the SSH key specified in the Secret. And also I have to specify Service Account for TaskRun.

**Comment by sbwsg** [comment]
Thanks a lot for taking the time to update the doc @OctoElla !

From my pov I don't think we can delete the existing auth content until the ServiceAccount-based approach is removed from Tekton, which is not something we have planned for yet I don't think. I will come back to this in a sec but first I also wanted to address the use of projected volumes :-

One issue with using `volumes` in Tasks and TaskRuns generally is that it requires the user to hard-code some volume configuration in their Task. A `volumeMount` must be added to a Step so that it receives a `volume`. For example, a user would need to take [the `git-clone` Task](https://github.com/tektoncd/catalog/blob/main/task/git-clone/0.5/git-clone.yaml) from the catalog and edit the `clone` Step to include a `volumeMount` with the name of the projected volume they define in their TaskRuns.

A different approach is for a Task to expose a Workspace for credentials. We see this with the `git-clone` Task, which defines an `ssh-directory` Workspace. The user is able to provide a Secret to this workspace in their TaskRun:

```yaml
kind: TaskRun
# ... skip some fields
spec:
  taskRef:
    name: git-clone
  workspaces:
  - name: ssh-directory
    secret:
      secretName: my-ssh-secret
```

This secret is structured just like an ssh dir, so can include a `id_rsa` entry and a `known_hosts` entry, but it doesn't need any of the Tekton-specific annotations or the specific type information.  It can just be a plain Opaque secret, the Workspace doesn't care.

The advantage is that the Task does not need to hard-code any `volumeMount` configuration. It just asks for a general workspace. Then the TaskRun fills in the details of exactly what kind of volume should be used for that workspace.

One down side is that Workspaces do not support projected volumes - this is definitely something we could add but it's not supported right now.

Coming back to the existing `auth.md` doc, I'd like to suggest a slightly different approach to take. Right now this doc only describes the ServiceAccount-based approach (also called `creds-init` in our codebase). What I'd like to suggest is that we add a new docs directory, `docs/auth/`, and then move the existing `auth.md` to `docs/auth/creds-init.md`. Then we could also add a `docs/auth/workspace-credentials.md` to describe the workspace-based approach and a `docs/auth/README.md` which can give a short explanation and links to the other docs. We can also add a `docs/auth/projected-volumes.md` describing that approach too (and maybe describing the pitfall that it requires editing the Task Steps?).

**Comment by afrittoli** [comment]
@OctoElla thank you for your contribution? Are you still pursuing this?

**Pull request closed without merging**